### PR TITLE
GlobalStyles: extract sanitize method

### PR DESCRIPTION
**Pull request by oandregal**
Extracted from #30541 so it can land in pieces and make reviews easier/quicker.

This PR extracts the sanitization to its own method and simplifies its logic so that changing the structure at #30541 is easier.

## How to test

`npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-test.php`

### CODE REVIEW

File: phpunit/class-wp-theme-json-test.php [hunk 1]
```diff
@@ -811,16 +809,10 @@ function test_get_template_parts() {
...
 
-		$this->assertEqualSetsWithIndex(
-			$template_parts,
-			array(
-				'small-header' => array(
-					'area' => 'header',
-				),
-			)
-		);
+		$this->assertEqualSetsWithIndex( $expected, $actual );
```
Comment (gziolo) [inline phpunit/class-wp-theme-json-test.php]: Nit: I have an observation, it's hard to follow when `$expected` and `$actual` are both extracted to their own variables.

File: lib/class-wp-theme-json.php [hunk 2]
```diff
@@ -485,12 +436,14 @@ private static function remove_keys_not_in_schema( $tree, $schema ) {
 		$tree = array_intersect_key( $tree, $schema );
 
 		foreach ( $schema as $key => $data ) {
-			if ( is_array( $schema[ $key ] ) && isset( $tree[ $key ] ) ) {
+			if ( is_array( $schema[ $key ] ) && is_array( $tree[ $key ] ) ) {
```
Comment (gziolo) [inline lib/class-wp-theme-json.php:447]: `isset` accounts also for the case when the `$key` in the `$tree` is not set. Is it safe to replace it with `is_array`?

File: phpunit/class-wp-theme-json-test.php (hunk 1 above)
Comment (oandregal) [reply phpunit/class-wp-theme-json-test.php]: Going to revert this change, it's not necessary for this to work.

What happened here was that the expected and actual values were inverted, hence the tests reported something I struggled to understand. Personally, I find it easier to have `expected`/`actual` variables for tests, as I can use them as anchors to navigate the rest of the code. I understand this is a style choice, though. :man_shrugging:

Comment (gziolo) [reply phpunit/class-wp-theme-json-test.php]: I'm probably biased by the tests I saw so far in WordPress core, like:

https://github.com/WordPress/wordpress-develop/blob/master/tests/phpunit/tests/blocks/block.php

File: lib/class-wp-theme-json.php (hunk 2 above)
Comment (oandregal) [reply lib/class-wp-theme-json.php:447]: It's not, good catch! Pushed another check above as well.

Would you know how to make unit tests fail if there are warnings? Somehow they're not reported for me when running `npm run test-unit-php`.

Comment (gziolo) [reply lib/class-wp-theme-json.php:447]: Is it possible that the env that runs phpunit has error reporting disabled?

File: lib/class-wp-theme-json.php [hunk 3]
```diff
@@ -295,75 +272,53 @@ class WP_Theme_JSON {
...
-				$styles_schema                                 = self::SCHEMA['styles'];
-				$this->theme_json['styles'][ $block_selector ] = self::remove_keys_not_in_schema(
-					$this->theme_json['styles'][ $block_selector ],
-					$styles_schema
-				);
+		$schema = array();
+		foreach ( $valid_block_names as $block_name ) {
+			$schema['styles'][ $block_name ]   = self::VALID_STYLES;
+			$schema['settings'][ $block_name ] = self::VALID_SETTINGS;
+		}
```
Comment (ockham) [inline lib/class-wp-theme-json.php:300]: For each `$block_name`, we're creating an identical copy of `self::VALID_STYLES` and `self::VALID_SETTINGS`, respectively. Would it be worth considering having just

```suggestion
		$schema['styles']   = self::VALID_STYLES;
		$schema['settings'] = self::VALID_SETTINGS;
```

and changing `remove_keys_not_in_schema()` accordingly?

**Review by ockham — APPROVED** [review]
Thanks @nosolosw! This is looking good and working well (judging by unit tests) 🎉 

I had [one note](https://github.com/WordPress/gutenberg/pull/30809#discussion_r613451531) about maybe simplifying some of the sanitization logic; but feel free to merge, we can discuss tackling this separately.

File: lib/class-wp-theme-json.php (hunk 3 above)
Comment (oandregal) [reply lib/class-wp-theme-json.php:300]: One of the struggles of this class has been how to make it adaptable to theme.json shape changes. The idea I explored in the latest rounds is to separate the logic that iterates over the tree from the logic to do something with a particular set of nodes. A visitor pattern of sorts. The hope is that, by centralizing the iteration logic, changes can be made easily in the future.

For example, in this case, `remove_keys_not_in_schema` is agnostic about the shape of theme.json, it just cares about removing nodes in the given tree that don't follow the given schema. The same logic applies to some other methods in this class after the latest changes landed in master: most of them become "visitors" of nodes, they know how to operate on a node or set of nodes (take styles out of them, process presets, etc) but are theme.json-shape agnostic. Then there're a couple of methods that are focused on building the paths to iterate over, so those are the only ones that need updating when theme.json changes.
